src/Server/RestRouterHandlers/SearchHandler.cpp

```diff
@@ -0,0 +1,219 @@
+            out->position() = out->buffer().begin();
+
+        writeString(jsonErrorResponse(e.message(), e.code()), *out);
+        writeChar('\n', *out);
+        out->next();
+        out->finalize();
+    }
+
+    return "";
+}
+
+String SearchHandler::getQuery(const Poco::JSON::Object::Ptr & payload) const
+{
+    /// Setup query settings
+    const auto & mode = payload->has("mode") ? payload->get("mode").toString() : "standard";
+
+    /// FIXME: to support 'realtime' mode in future
+    /// FIXME: change the behavior of 'verbose' mode. Currently in frontier service, 'standard' and 'verbose' mode are the same
+    if (mode == "standard" || mode.empty() || mode == "verbose")
+        query_context->setDefaultFormat("JSONCompactEachRowWithNamesAndTypes");
+
+    const auto & start_time = payload->has("start_time") ? payload->get("start_time").toString() : "";
+    if (!start_time.empty())
+        query_context->setTimeParamStart(start_time);
+
+    const auto & end_time = payload->has("end_time") ? payload->get("end_time").toString() : "";
+    if (!end_time.empty())
+        query_context->setTimeParamEnd(end_time);
+
+    query_context->setSetting("unnest_subqueries", true);
+    query_context->setSetting("asterisk_include_materialized_columns", true);
+    query_context->setSetting("asterisk_include_alias_columns", true);
+
+    std::vector<String> query_parts;
+    query_parts.push_back(fmt::format("SELECT * FROM ({})", payload->get("query").toString()));
+    if (payload->has("offset") && payload->has("page_size"))
+    {
+        int offset = payload->getValue<int>("offset");
+        int page_size = payload->getValue<int>("page_size");
+        if (page_size > 0 && offset >= 0)
+            query_parts.push_back(fmt::format("LIMIT {}, {}", offset, page_size));
+    }
+    return boost::algorithm::join(query_parts, " ");
+}
+
+bool SearchHandler::validatePayload(const Poco::JSON::Object::Ptr & payload, String & error)
+{
+    error.clear();
+    if (!validateSchema(search_schema, payload, error))
+        return false;
+
+    if (payload->has("mode"))
+    {
+        const auto & mode = payload->getValue<String>("mode");
+        if (mode != "standard" && mode != "verbose")
+        {
+            error = fmt::format("Invalid 'mode': {}, only support 'standard', 'verbose'", mode);
+            return false;
+        }
+    }
+
+    bool has_limit = payload->has("offset");
+    bool has_page_size = payload->has("page_size");
+    if (has_limit && has_page_size)
+    {
+        if (payload->getValue<int>("offset") < 0 || payload->getValue<int>("page_size") <= 0)
+        {
+            error = "Invalid 'limit' or 'page_size";
+            return false;
+        }
+    }
+    else if (has_limit || has_page_size)
+    {
+        error = "Missing 'limit' or 'page_size";
+        return false;
+    }
+
+    return true;
+}
+
+bool SearchHandler::validateQuery(const String & query, String & error) const
+{
+    error.clear();
```

Comment:
we don't need clear

Reply:
I think run clear is better to avoid the impact from last call with non empty error message.